apollo-compiler/src/main/kotlin/com/apollostack/compiler/Util.kt

```diff
@@ -1,10 +1,49 @@
 package com.apollostack.compiler
 
-import com.squareup.javapoet.AnnotationSpec
+import com.squareup.javapoet.*
 import javax.annotation.Nullable
+import javax.lang.model.element.Modifier
 
 fun String.normalizeTypeName() = removeSuffix("!").removeSurrounding("[", "]").removeSuffix("!")
 
+fun TypeName.overrideTypeName(typeNameOverrideMap: Map<String, String>): TypeName {
+  if (this is ParameterizedTypeName) {
+    val typeArguments = typeArguments.map { it.overrideTypeName(typeNameOverrideMap) }.toTypedArray()
+    return ParameterizedTypeName.get(rawType, *typeArguments)
+  } else if (this is ClassName) {
+    return ClassName.get(packageName(), typeNameOverrideMap[simpleName()] ?: simpleName())
+  } else {
+    return this
+  }
+}
+
+fun MethodSpec.overrideMethodReturnType(typeNameOverrideMap: Map<String, String>) =
+    MethodSpec.methodBuilder(name)
+        .returns(returnType.overrideTypeName(typeNameOverrideMap).annotated(returnType.annotations))
+        .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
+        .build()
+
+fun TypeSpec.resolveNestedTypeNameDuplication(usedOuterTypeNames: List<String>): TypeSpec {
+  val usedTypeNames = usedOuterTypeNames + name
+  val typeNameOverrideMap = typeSpecs.map {
```

Comment:
you could also do `typeSpecs.map { it.name }.map { typeName ->` since it looks like you just need the name